src/proj.rs

```diff
@@ -0,0 +1,791 @@
+use std::ops::Mul;
+use rayon::prelude::*;
+
+#[derive(Copy, Clone, Debug)]
+enum TransformationClass {
+    Translation,
+    Affine,
+    Projection
+}
+
+/// A 2d Projective transformation, stored as a row major 3x3 matrix.
+/// Transformations combine by multiplication.
+///
+/// I.e. to define rotation around image center
+/// ```
+/// use imageproc::proj::*;
+/// let (cx,cy) = (320.0,240.0);
+/// let c_rotation = Proj::trans(cx,cy)*Proj::rot(30.0)*Proj::trans(-cx, -cy);
+/// ```
+#[derive(Copy, Clone, Debug)]
+pub struct Proj {
+    transform: [f32; 9],
+    inverse:   [f32; 9],
+    class: TransformationClass,
+}
+
+impl Proj {
+    /// Create a 2d projective transform from a row-major 3x3 matrix in homogeneous coordinates.
+    /// Matrix must be invertible, otherwise it does not define a Projection (by definition).
+    pub fn from_matrix(transform: [f32; 9]) -> Option<Proj> {
+        let transform = normalize(transform);
+        let class = class_from_matrix(transform);
+        try_inverse(&transform)
+            .map(|inverse| 
+                 Proj { transform, inverse, class })
+    }
+
+    /// Defines a translation by (tx, ty)
+    pub fn trans(tx: f32, ty: f32) -> Proj {
+        Proj { 
+            transform: [
+                1.0, 0.0, tx,
+                0.0, 1.0, ty,
+                0.0, 0.0, 1.0],
+            inverse: [
+                1.0, 0.0, -tx,
+                0.0, 1.0, -ty,
+                0.0, 0.0, 1.0],
+            class: TransformationClass::Translation,
+        }
+    }
+
+    /// Defines a rotation around the origin by angle theta degrees. Origin is (0,0) pixel
+    /// coordinate, usually top left corner. To rotate around image center combine rotation with 
+    /// two translations: T*Rotation*T^-1.
+    pub fn rot(theta: f32) -> Proj {
+        let theta = theta.to_radians();
+        let s = theta.sin();
+        let c = theta.cos();
+        Proj { 
+            transform: [
+                c,  -s,   0.0,
+                s,   c,   0.0,
+                0.0, 0.0, 1.0],
+            inverse: [
+                c,   s,   0.0,
+                -s,  c,   0.0,
+                0.0, 0.0, 1.0],
+            class: TransformationClass::Affine,
+        }
+    }
+
+    /// Creates an anisotropic scaling (sx,sy).
+    pub fn scale(sx: f32, sy: f32) -> Proj {
+        Proj { 
+            transform: [
+                sx,  0.0, 0.0,
+                0.0, sy,  0.0,
+                0.0, 0.0, 1.0],
+            inverse: [
+                1.0/sx, 0.0, 0.0,
+                0.0, 1.0/sy, 0.0,
+                0.0, 0.0, 1.0],
+            class: TransformationClass::Affine,
+        }
+    }
+
+    /// Inverts the transformation.
+    pub fn inv(self) -> Proj {
+        Proj { transform: self.inverse, inverse: self.transform, class: self.class }
+    }
+
+    /// Calculates a homography from a set of four control point pairs
+    pub fn from_cp(from: [(f32, f32); 4], to: [(f32, f32); 4]) -> Option<Proj> {
```

Comment:
This abbreviation seems unnecessarily cryptic. `from_control_points`?

Reply:
I agree. Will do in my next update.

Reply:
Done.